### ISSUE TITLE
Azure Portal: Cannot access "Monitor" as "Monitoring Reader"

### ISSUE THREAD

**Issue by ThomasVandenbon** [Azure/Azure-Functions#1882]
_My apologies if I'm filing this issue in the wrong repository, however I couldn't find a repository where I could post issues for Azure Functions in the Azure Portal._

I've been granted the following pre-made RBAC roles on a Function App:
| Role | Description | Scope |
| ----- | ------------ | ------- |
| Monitoring Reader | Can read all monitoring data | Resource group (Inherited) |
| Reader | View all resources but does not allow you to make any changes | Resource group (Inherited) |

When I navigate to "Functions" in the Function App and select a specific Azure Function, I cannot access the "Monitor" section of the Function. When I hover over "Monitor" it says "This feature requires write permission to your app".

This seems like a bug, as I'd expect to be able to read monitoring data as a "Monitoring Reader", since it says "Can read all monitoring data".

_Note: as this is a production environment, I do not want to have write permissions as the message suggests._

**Comment by v-bbalaiagar** [comment]
Hi @ThomasVandenbon, More than a bug, its a role assignment issue and requires a support ticket for resolution. Please follow the below path to raise a support ticket- Portal > menu > help and support >New support request

**Comment by ThomasVandenbon** [comment]
@v-bbalaiagar I was hopeful when you wrote that this is more than a bug.

We opened a support request and have since received some answers.

1. The first answer suggested that we had failed to correctly apply the Monitoring Reader / Reader roles and guided us on how to do this. This was however not the case and the roles were correctly set.

2. The second answer then confirmed the issue (after reproducing it on the side of Microsoft) and escalated it to a Technical Advisor.

3. The final answer then informed us that this is by design 😒 telling us to go vote for a fix here:
https://feedback.azure.com/forums/355860-azure-functions/suggestions/35024440-enable-reader-access-to-view-azure-functions-monit

What baffles me, is that every person I've spoken to has instantly identified it as a bug, yet there seems to be no inclination to fix this in the near future. 🤷‍♂️

Also, if it was that easy to do via Application Insights, then why would there even be a need for this screen.  
It seems I won't be monitoring my Azure Functions in the near future...

**Comment by denalisk** [comment]
Bumping this because I'm confused about what the necessary role is to access "Monitor" for azure functions. I'm an owner at the subscription level, and yet the "monitor" blade is greyed out for me in the Azure functions project, with the prompt "this feature requires write permission to your app". 

I had access to the Monitor blade as recently as a month ago, and to my knowledge no role changes have happened. Are there new role requirements/workarounds that I need to access the monitor blade?

N.B. it looks like there are ongoing Azure AD issues, at time of writing this, so maybe that's the real issue

**Comment by v-bbalaiagar** [comment]
Hi @denalisk, Thank you for your feedback! Checking internally on this issue and update you with the findings.

**Comment by denalisk** [comment]
@v-bbalaiagar it does look like my issue was likely connected to the AD issues at the time, I have since been able to access the monitor blade in my function resource

**Comment by v-bbalaiagar** [comment]
Hi @ThomasVandenbon, Closing this issue as this issue hasn't re-occurred. Please feel free to reach back to us in case of any further queries related to this issue.

**Comment by ThomasVandenbon** [comment]
@v-bbalaiagar, I don't understand what you mean by not re-occuring?

As far as I know, nothing has changed.
Which means that a user with the "Monitoring reader" and/or "Reader" role cannot use the "Monitor" feature of Azure Functions.

**Comment by v-bbalaiagar** [comment]
Hi @ThomasVandenbon, Apologies for the confusion, I shall discuss this internally and get back to you with the findings.

**Comment by v-bbalaiagar** [comment]
Hi @ehamai, Can you kindly look into this issue.

**Comment by estokes-vs** [comment]
@v-bbalaiagar @ehamai any update on this issue?  a user looking to read logs, or data under the `monitor` tab should not be required to have contributor access and be able to delete the items

**Comment by fslefevre78** [comment]
any update? I am facing the same issue, the Azure Function Monitoring tab is grayed-out however correct RBAC roles are applied.

**Comment by sajikag** [comment]
It worked after adding contributor access to related app-insights instance

**Comment by TroyWitthoeft** [comment]
Why even have the reader or monitoring reader roles if it takes contributor to read the logs?  Yes, I know App Insights is an option, but for what I want to share, succinct success/failures and log messages, the monitoring tab is perfect.  App Insights and KUSTO can be a bit overwhelming for new folks.  

Wouldn't it be both **more intuitive** and **more secure** (least-privilege principles) if we didn't have to grant contributor level access in order to read logs?

**Comment by ehamai** [comment]
Sorry I wasn't tracking this thread.  The reason why you need contributor access is because the ikey/connection string for your AI resource is stored as an app setting which can contain secrets.  If we wanted to remove the contributor requirement, we'd need the runtime itself to support it as something other than an app setting.  @fabiocav I'll let you respond, but unless someone from the runtime decides to change this dependency, then this isn't something that can be fixed from the portal itself.

**Comment by mfreedm52** [comment]
It doesn't look like there is any secrets exposed via the monitor panel, so I don't see why the monitor reader role should be restricted from in. What is the monitor reader role for if not the monitor panel?

**Comment by TroyWitthoeft** [comment]
Agreed.  Not allowing the "monitoring reader" role to "read the monitor" is seems pretty unintuitive.  Also, I don't understand why only "contributors" can read secrets.   If "reading secrets" is a permission then can we factor that out and grant it to both the "contributor" and "monitoring reader" roles?  Or - because I know editing the permissions of established roles is a big deal - is there some way we can improve this?  Such as having the portal detect that the visitor has monitoring reader but not contributor and alert or link them over to AppInsights?  Anything would help.   

Devs link over to those run logs because they are succinct and useful.  They want to share logs with others.  Those others are frustrated when they can't view those pages without getting edit rights on the object.  How can we lessen the friction here?

**Comment by Beelzenef** [comment]
I suppose we can create a custom role to be able to see Monitor... which of the [actions listed in the documentation](https://docs.microsoft.com/en-us/azure/azure-monitor/roles-permissions-security#monitoring-permissions-and-azure-custom-roles) is appropriate? Or is it a different one that is not listed?

It is not the best solution, but it is a possibility with the inconsistency of this topic. There is no point in having that role and not being able to see the Function Apps monitor.

**Comment by balag0** [comment]
@ehamai Just to clarify your earlier comment, 
>If we wanted to remove the contributor requirement, we'd need the runtime itself to support it as something other than an app setting

Is this referring to the functions runtime? Are these appsettings read via arm? Or is there some api that exposed by functions runtime?

**Comment by Beelzenef** [comment]
We tried to create a custom role, [following the documentation](https://docs.microsoft.com/en-us/azure/azure-monitor/roles-permissions-security#monitoring-permissions-and-azure-custom-roles)... and we have no progress on this issue

![image](https://user-images.githubusercontent.com/6389665/149739917-476e9e0b-b98b-42d4-8fe3-c2751a3732ea.png)

**Comment by abirstolov** [comment]
Any updates?

**Comment by Beelzenef** [comment]
any news, @v-bbalaiagar @ehamai?

**Comment by TroyWitthoeft** [comment]
*sigh* I guess I'll pile on.  Any updates?

**Comment by MIG0Z** [comment]
Same problem here... any news on the subject? @ehamai

**Comment by Annesars90** [comment]
Is it possible to list which permissions are needed to be able to use Monitor in Function Apps when we don't want to assign Contributor, so we can create a custom role?

**Comment by IanMoroney** [comment]
I actually have contributor permission at the subscription level, and *still* can't get this to work.

**Comment by IanMoroney** [comment]
looking at it again, it's definitely buggy.
I refreshed the azure portal, and it was fine, i could access it.

**Comment by komayama** [comment]
I added Website contributor role that azure provide build-in roles and there is not gray-out on "monitor" screen of azure portal. Just for your information.

https://docs.microsoft.com/en-us/azure/role-based-access-control/built-in-roles#website-contributor

**Comment by Beelzenef** [comment]
> I added Website contributor role that azure provide build-in roles and there is not gray-out on "monitor" screen of azure portal. Just for your information.
> 
> https://docs.microsoft.com/en-us/azure/role-based-access-control/built-in-roles#website-contributor

so... is this a workaround?

**Comment by bjoljo** [comment]
When using a custom role with the permission "Microsoft.Web/sites/Write" monitor is not greyed-out.

The description for the permission says "Create a new Web App or update an existing one" which might not be  an acceptable level of permission for a user whos job is to "monitor" the function.

Have any one else found a permission that enables "monitor" with out giving the user some sort of contributor role?

**Comment by tokidoki11** [comment]
can you help us by giving what permission really needed at minimum to read the log only?
thus we can create custom role for this one at least for the time being

@ehamai

**Comment by AshutoshNirkhe** [comment]
I thought only I was going crazy while trying to grant our SRE with Monitor Read only access to know that it didn't work for him. But looks like its a real issue. 
Is granting "Microsoft.Web/sites/Write" (which at least looks better than Contributor access) the only workaround for now ?

**Comment by tokidoki11** [comment]
Nah, tried giving the sites write plus the monitor reader, it doesn't work

It needs something.
Havent contacted the support yet

**Comment by AshutoshNirkhe** [comment]
> Nah, tried giving the sites write plus the monitor reader, it doesn't work
> 
> It needs something. Havent contacted the support yet

Oh ok, thanks for the update!

**Comment by UltraGC** [comment]
I am running into this situation as well.  It appears we need to grant write access to app in order to view the function app's monitor info.  This, to me, is a bad workaround.  For my need, I only want to grant our monitoring team read access.  Granting them write access is not an option.

**Comment by iqoOopi** [comment]
Same issue here. We granted "Monitor contributor" role but it is still greyed out.....

**Comment by IanKemp** [comment]
@fabiocav Being able to monitor functions and view their logs is the absolute, fundamental, most basic requirement that makes the difference between a decision to use Azure Functions and use something like AWS Lambda.

The fact that this was **ever** an issue, never mind is **still** an issue _more than a year after being reported_, implies you literally did not ever test that the "reader" role could perform this functionality, which isn't just poor; it's unacceptable for a product that customers pay money for. Once again the Azure Functions team has failed a company I work at.

FIX THIS.

**Comment by DCPDamon** [comment]
I've had a case open for a month, and I've lost count of the emails back and forth with support. I'm told they're escalating my case to so I can talk to a tech lead about it next week.  I did a ton of trial and error and dialed in what I believe is the current minimum required permissions.  You need Reader on the Function App and Application Insights, and you need Microsoft.Web/sites/Write on the Function App.

A few notes based on previous comments.  Monitoring Reader has to be granted on the Application Insights resource, not the Function App.  Or, you can grant Reader on Application Insights, that still works and Reader is less permissive than Monitoring Reader.

Microsoft.Web/sites/Write can be found in Contributor or Website Contributor role.  Or, you can make a custom role, as was mentioned above.

I also don't understand the comment around needing Contributor so you can read secrets in App Settings.  I have an account that has Reader and Microsoft.Web/sites/Write via a custom role.  It can use the Monitoring blade, but it cannot see App Settings in Portal.  Is that proof against the statement or am I missing something? Also, you don't need Contributor to see App Settings, that privilege is controlled solely by the Microsoft.Web/sites/config/list/Action permission (which I've alo proven out by trial and error). 

Wish me luck with support 🤞.

**Comment by m-adami** [comment]
This is still an issue almost two years after the issue was opened. This is a big deal for everybody who wants to implement a decent CI/CD flow where developers don't have direct write access to functions but still need to be able to read logs. I can't stress enough how basic I consider this kind of access, to be honest. Having to grant write access to functions just to read logs it's just blowing my mind.

**Comment by albedosehen** [comment]
Ran into this issue today and we currently have Monitor Contributor assigned. Is this being addressed by Microsoft?

**Comment by DCPDamon** [comment]
I talked to product support in May 2022. Their "resolution" was:

> •	This is a known issue and is been addressed internally. However, there is currently no ETA for a fix
> •	As a workaround, users with "reader" permission can access App Insight directly for application loggings and metrics

So, I'm guessing they're not working on it.

**Comment by ifatssane** [comment]
I am facing the same issue.

**Comment by ranrsg** [comment]
Same here

**Comment by shogd3838** [comment]
Hello it looks like this issue is still not solved.

**Comment by Beelzenef** [comment]
> I talked to product support in May 2022. Their "resolution" was:
> 
> > •	This is a known issue and is been addressed internally. However, there is currently no ETA for a fix
> > •	As a workaround, users with "reader" permission can access App Insight directly for application loggings and metrics
> 
> So, I'm guessing they're not working on it.

May 2023, still working on it (I guess)

**Comment by jlcfg** [comment]
I just ran into the same issue. I ended up giving QA contributor access 👯

**Comment by sahangunathilaka** [comment]
It worked when added contributor access to both app insight service and Azure function.

**Comment by AnKiLa** [comment]
This issue is very annoying. We have got a strictly locked down productive environment and dev team has only got Monitoring Contributor role assigned. Currently it's not possible to monitor azure functions functionality for them. Because of security reasons they will not get more priviledges roles for just look up some logs. A fix would be greatly appreciated.

**Comment by TroyWitthoeft** [comment]
A pitch and plea for this workaround/mitigation again

> having the portal detect that the current page visitor has monitoring reader but not contributor, and then popup a link that directs them over to AppInsights? Preferably with the KQL prefilled with the same query that is feeding the monitor tab? Anything would help.

**Comment by jj-obrien** [comment]
This continues to be an issue two years on. We sorely need the function app monitoring tab to be decoupled from `Microsoft.Web/sites/Write` permission - it grants too wide an access for people who just want to troubleshoot simple invocation issues.

**Comment by Kaloszer** [comment]
Get a move on, it's been OVER 2 years.

**Comment by Kaloszer** [comment]
Hello 2024! Still an issue

**Comment by lettucemode** [comment]
In case it helps others, here's an App Insights query similar to what you get from the Monitor page, just in flattened rows:

```
requests
| join (
    traces 
    | union exceptions
) on operation_Id
| project
    timestamp,
    // id,
    operation_Name,
    success,
    resultCode,
    duration,
    operation_Id,
    cloud_RoleName,
    // invocationId=customDimensions['InvocationId'],
    message = iff(message != '', message, iff(innermostMessage != '', innermostMessage, customDimensions.['prop__{OriginalFormat}']))
| where timestamp > ago(30d)
| where cloud_RoleName =~ '<your function app name>' and operation_Name =~ '<your function name>'
| order by timestamp desc
| take 20
```

This issue also bit us so hopefully some action is taken soon. We'll probably be asking our devs to connect to the relevant app insights instance and run this instead of using the Monitor blade in the meantime.

**Comment by InsomniumBR** [comment]
Thanks @lettucemode ! I did some changes to create one that groups by invocation. My use case is a time triggered function, so that helps to understand the flow of the function took on each invocation.

```
requests
| join (
    traces 
    | union exceptions
) on operation_Id
| where timestamp > ago(1d)
| where cloud_RoleName =~ '<your function app name>' and operation_Name =~ '<your function name>'
| summarize 
    TimeStamp = min(timestamp1),
    LogMessages = make_list(
        strcat(tostring(timestamp1), " - ", iff(message != '', message, iff(innermostMessage != '', innermostMessage, customDimensions.['prop__{OriginalFormat}'])))
    )
by operation_Id
| order by TimeStamp desc
| take 200
| project
    Invocation = operation_Id,
    TimeStamp,
    LogMessages
```

Sample view:

![image](https://github.com/Azure/Azure-Functions/assets/2073799/ec718c6a-56fc-4a50-94c6-6931eb83a8bb)

**Comment by Shum1905** [comment]
FYI same thing happened here on App Service with permission at the subscription level. In my case, I was not able to access "Quota" and "Change App Service plan" tabs too. But after reloading several times, it appeared to be no problem and was able to access all tabs.
It is definetely a bug and needs improvements...

**Comment by RicardoLans** [comment]
Literally almost 3 years further and still no progress or any update on this issue? Do we even exist?

**Comment by jonavask** [comment]
Echoing the above. Applying `microsoft.web/sites/write` permission just so you get the ability to view the monitor page is not good practise!

**Comment by rreeves8** [comment]
Bumping this again in 2025 lol

**Comment by sam-bradshaw-wcmc** [comment]
Same issue again

**Comment by IanKemp** [comment]
> [@fabiocav](https://github.com/fabiocav) Being able to monitor functions and view their logs is the absolute, fundamental, most basic requirement that makes the difference between a decision to use Azure Functions and use something like AWS Lambda.
> 
> The fact that this was **ever** an issue, never mind is **still** an issue _more than a year after being reported_, implies you literally did not ever test that the "reader" role could perform this functionality, which isn't just poor; it's unacceptable for a product that customers pay money for. Once again the Azure Functions team has failed a company I work at.
> 
> FIX THIS.

3 years later and MS still hasn't fixed this, guess shoving "AI" into everything is more important than making actual useful products. What a pathetic state of affairs.

**Comment by Kaloszer** [comment]
@IanKemp banger

**Comment by TroyWitthoeft** [comment]
Hey CoPilot

### You are an expert-level coder who cares deeply about customer feedback. 
### Customers are frustrated that Monitoring Reader and Reader roles can’t see the Monitor tab. 
### Update the authorization so they can actually monitor their Functions. 
### Bonus points if you make it look like you cared about this all along.

**Comment by AbdelHajou** [comment]
Can we get clarity on when this issue will be prioritized?
This bug is forcing us to assign Contributor level permissions to users just to view monitoring data for their Function Apps. That’s a significant security and governance concern, and it's impacting our ability to manage access appropriately.
